### ISSUE TITLE
`no-useless-undefined` with `strictNullCheck` and `<T | undefined>`

### ISSUE THREAD

**Issue by castral** [sindresorhus/eslint-plugin-unicorn#2899]
When you have `strictNullChecks` enabled in tsconfig, you are explicitly required to pass `undefined` as a value to a function when it is not optional and a generic or the parameter type for that argument is a union with `undefined`, for example:

```ts
function test<T extends object | undefined>(arg: T): void {

}

test({}); // fine
test(undefined); // fine, but eslint complains on rule unicorn/no-useless-undefined
test(); // not fine, compiler throws TS2554: Expected 1 arguments, but got 0
```

I ran into this issue with the `unicorn/no-useless-undefined` rule and worked around it by setting `checkArguments: false` but it seems like an incorrect implementation to me.

**Comment by castral** [comment]
That doesn't really make a lot of sense because in this case it was a third party library that was the call-site: RxJS. And the `undefined` was a valid usage in the union, it could genuinely be an undefined, uninitialized value. So the rule is incorrect because there are valid corner cases where it throws errors.

**Comment by fregante** [comment]
Reopening to discuss the removal of the rule **for TS files** because TS should already be able to tell the user whether `undefined` is useless or required like in this case. 

A new rule could then be devised to cover the *definition* of functions that "accept useless `undefined` types"
